oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/CompressedDocumentPropertyState.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+    /**
+     * Read a {@code PropertyState} from a {@link JsopReader}.
+     *
+     * @param name the name of the property state
+     * @param store the store
+     * @param reader the reader
+     * @return new property state
+     */
+    static PropertyState readProperty(String name, DocumentNodeStore store, JsopReader reader) {
+        if (reader.matches(JsopReader.NUMBER)) {
+            String number = reader.getToken();
+            Long maybeLong = LongUtils.tryParse(number);
+            if (maybeLong == null) {
+                return new DoublePropertyState(name, Double.parseDouble(number));
+            } else {
+                return new LongPropertyState(name, maybeLong);
+            }
+        } else if (reader.matches(JsopReader.TRUE)) {
+            return BooleanPropertyState.booleanProperty(name, true);
+        } else if (reader.matches(JsopReader.FALSE)) {
+            return BooleanPropertyState.booleanProperty(name, false);
+        } else if (reader.matches(JsopReader.STRING)) {
+            String jsonString = reader.getToken();
+            if (jsonString.startsWith(TypeCodes.EMPTY_ARRAY)) {
+                int type = PropertyType.valueFromName(jsonString.substring(TypeCodes.EMPTY_ARRAY.length()));
+                return PropertyStates.createProperty(name, emptyList(), Type.fromTag(type, true));
+            }
+            int split = TypeCodes.split(jsonString);
+            if (split != -1) {
+                int type = TypeCodes.decodeType(split, jsonString);
+                String value = TypeCodes.decodeName(split, jsonString);
+                if (type == PropertyType.BINARY) {
+
+                    return  BinaryPropertyState.binaryProperty(name, store.getBlobFromBlobId(value));
+                } else {
+                    return createProperty(name, StringCache.get(value), type);
+                }
+            } else {
+                return StringPropertyState.stringProperty(name, StringCache.get(jsonString));
+            }
+        } else {
+            throw new IllegalArgumentException("Unexpected token: " + reader.getToken());
+        }
+    }
+
+    /**
+     * Read a multi valued {@code PropertyState} from a {@link JsopReader}.
+     *
+     * @param name the name of the property state
+     * @param reader the reader
+     * @return new property state
+     */
+    PropertyState readArrayProperty(String name, JsopReader reader) {
+        return readArrayProperty(name, store, reader);
+    }
+
+    /**
+     * Read a multi valued {@code PropertyState} from a {@link JsopReader}.
+     *
+     * @param name the name of the property state
+     * @param store the store
+     * @param reader the reader
+     * @return new property state
+     */
+    static PropertyState readArrayProperty(String name, DocumentNodeStore store, JsopReader reader) {
+        int type = PropertyType.STRING;
+        List<Object> values = Lists.newArrayList();
```

Comment:
I think this can be refactored too, and then entirely avoid the import of the guava dependency. I think that's what @reschke wanted to point out.

Reply:
Done. It wasn't very clear from the comment.